### PR TITLE
Private data member renaming: use underscore as name suffix

### PR DESCRIPTION
**Pull request by wenqing**
As tilted.

The fixes #2762.

Do we need name suffix or prefix? The trailing underscore makes name really ugly. Any modern editor can highlight the data members.

### CODE REVIEW

**Comment by wenqing** [comment]
If everybody agrees that  trailing underscore naming is good. Please review once the test becomes green. If you like not to use prefix or suffix in the data member naming, I will make the removal of the underscore.

**Comment by endJunction** [comment]
Just curious? How do you distinguish between private and non-private members?

**Comment by wenqing** [comment]
> Just curious? How do you distinguish between private and non-private members?

by just checking the names headed by underscore.

**Comment by endJunction** [comment]
I recommend you to split the changes into several small peaces, otherwise it is difficult to keep up with the current changes and fixing many many remaining issues.
I'd recommend you to use clang-tidy for this operation. Sed is nice chap but he does not know c++ (unfortunately).

**Comment by bilke** [comment]
A trailing underscore is ok for private members for me.

**Comment by wenqing** [comment]
> I recommend you to split the changes into several small peaces, otherwise it is difficult to keep up with the current changes and fixing many many remaining issues.
> I'd recommend you to use clang-tidy for this operation. Sed is nice chap but he does not know c++ (unfortunately).

sed works well for member naming  change. The only thing is that it moves the underscore in quot, like "_foo", which mostly appears in the strings of file paths for IO or for system call in the source code. I will fix it later in this afternoon.

**Comment by wenqing** [comment]
Fixed the underscore in quot issue in strings, which makes benchmarks run on eve. Next: recovery some source code documentations.

**Comment by wenqing** [comment]
Source code documentation is OK now. All tests are green before checking the source code documentation. I am sure that
* all the private data member names were changed properly.
* The variable names, which have underscore inside the name string, were not touched. 
* The only exception is in EigenTools.h, where the underscore is used as the prefix of the function argument name. These name were changed as well. See:
https://github.com/ufz/ogs/pull/2962/commits/3622a05376ff64adeab1de418adce2663a595130#diff-bebcda64ace03d7ec81a84a7971863b8
   I don't know why these function arguments were named in that way. If there is no special meanings for that naming, I can remove that prefix from those function argument names.

Based on about points, to review the PR by every piece of the changes is not necessary. If the utility  tools are tested in ctest, the only thing to check is GUI/data explorer.  But I am sure that the utility  tools and GUI/data explorer should work.

File: Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp [hunk 1]
```diff
@@ -69,7 +69,7 @@ int main (int argc, char* argv[])
     std::string const name =
         BaseLib::extractBaseNameWithoutExtension(mesh_arg.getValue());
     // create file
-    std::string const new_matname(name + "prop_");
```
Comment (endJunction) [inline Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: these must not be changed. This is not a single occurrence. You will have to update your scripts.

Comment (wenqing) [reply Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: It is gone with my fixing of this afternoon. I think you forgot to fetch and update.

Comment (endJunction) [reply Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: Sorry. My bad.

File: Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ int main (int argc, char* argv[])
...
+    std::string const new_matname(name + "_prop");
```
Comment (endJunction) [inline Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: Sorry to reopen it again, but it seems this change is not good.

Comment (wenqing) [reply Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: I see that the correction was committed this commit for Process.  It should be committed to 76761fef, which is for Application, where the error was introduced. Anyway, this error is gone, and the file was not touched.

Comment (endJunction) [reply Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: Strange. I still see it in your commits. Also if I run git blame locally (or [view](https://github.com/wenqing/ogs/blame/data_member_naming/Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp#L72) on github for that matter) the change is there. You might need to fix the commits and push again.

Comment (wenqing) [reply Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: It was changed in the commit for  `Application`, and it was changed back again in the commit for `ProcessLib` (should be changed back in `Application`). Now the change of the file was dropped in the commit for  `Application`, and it dose not appear in commits any more.

Comment (wenqing) [reply Applications/Utils/FileConverter/generateMatPropsFromMatID.cpp]: The implementation in Utils (including file names and paths) are tested with unit test.

**Comment by wenqing** [comment]
@endJunction Thanks.

File: ChemistryLib/PhreeqcIOData/Output.h [hunk 3]
```diff
@@ -24,7 +24,7 @@ class BasicOutputSetups
 public:
     explicit BasicOutputSetups(std::string const& project_file_name,
                                bool const use_high_precision_)
-        : output_file(project_file_name + "_phreeqc.out"),
```
Comment (endJunction) [inline ChemistryLib/PhreeqcIOData/Output.h:27]: This must not be changed.

Comment (wenqing) [reply ChemistryLib/PhreeqcIOData/Output.h]: Checked by 
1. finding all strings and dump the results to file by the command of
```
 grep -R '"'  -r -i --include \*.h --include \*.cpp > comp.txt
```
2. comparing the file by ufz/master and the branch of this PR. In addition to the one you found, there is another in PhreeqcIO.cpp.  These two changes were just dropped.

File: ProcessLib/SurfaceFlux/SurfaceFluxData.h [hunk 4]
```diff
@@ -103,7 +103,7 @@ struct SurfaceFluxData
         // TODO (TomFischer) output, if output classes are ready this has to be
         // changed
         std::string const fname =
-            BaseLib::dropFileExtension(output_mesh_file_name) + "t__" +
+            BaseLib::dropFileExtension(output_mesh_file_name) + "_t_" +
```
Comment (endJunction) [inline ProcessLib/SurfaceFlux/SurfaceFluxData.h]: I thought this changes were removed already, weren't they?

File: ProcessLib/TES/TESReactionAdaptor.cpp [hunk 5]
```diff
@@ -52,138 +52,138 @@ std::unique_ptr<TESFEMReactionAdaptor> TESFEMReactionAdaptor::newInstance(
 
 TESFEMReactionAdaptorAdsorption::TESFEMReactionAdaptorAdsorption(
     TESLocalAssemblerData const& data)
-    // caution fragile: this relies in this constructor b eing called __after__
```
Comment (endJunction) [inline ProcessLib/TES/TESReactionAdaptor.cpp:55]: Comment's, not containing private member references also should not be changed.

File: ProcessLib/SurfaceFlux/SurfaceFluxData.h (hunk 4 above)
Comment (wenqing) [reply ProcessLib/SurfaceFlux/SurfaceFluxData.h]: It was changed the commit for the header files of ProcessLib and changed back in the commit for the cpp files of ProcessLIb. At the end, the green one is taken. If you look at the checkout of the branch, that line is kept as the one in ufz/master.

File: ProcessLib/TES/TESReactionAdaptor.cpp (hunk 5 above)
Comment (wenqing) [reply ProcessLib/TES/TESReactionAdaptor.cpp:55]: Recovered.

File: ProcessLib/SurfaceFlux/SurfaceFluxData.h (hunk 4 above)
Comment (wenqing) [reply ProcessLib/SurfaceFlux/SurfaceFluxData.h]: Moved the change-back to the commit for the header files of ProcessLib.

File: ProcessLib/Output/Output.h [hunk 6]
```diff
@@ -87,32 +87,32 @@ class Output
...
-    std::multimap<Process const*, MeshLib::IO::PVDFile> _process_to_pvd_file;
+    std::multimap<Process const*, MeshLib::IO::PVDFile> process_to_pvd_file_;
 
     /**
      * Get the address of a PVDFile from corresponding to the given process.
-     * @param process    Process.
+     * @param process Process.
      * @param process_id Process ID.
-     * @param mesh_name_for_output mesh name for the output.
+     * @param mesh_name_for_output Mesh file name for output
```
Comment (endJunction) [inline ProcessLib/Output/Output.h:115]: Still, there are unrelated changes in the comments.

**Comment by endJunction** [comment]
And also there are changes of the macro names.

File: ProcessLib/Output/Output.h (hunk 6 above)
Comment (wenqing) [reply ProcessLib/Output/Output.h]: > Still, there are unrelated changes in the comments.

keen eye! Thanks.
Forget to run grep for checking after rebase. After ran grep '#ifdef' and '#if defined'. The follwing macro definitions were corrected
```
Applications/DataExplorer/mainwindow.cpp:#ifdef WIN32_
GeoLib/Polygon.cpp:#if _GNUC___ <= 4 && (_GNUC_MINOR___ < 9)
GeoLib/Polygon.h:#if _GNUC___ <= 4 && (_GNUC_MINOR___ < 9)
```


Still, there are unrelated changes in the comments.
MathLib::Point3d getBulkElementPoint(MeshLib::Quad const& quad,

Comment (wenqing) [reply ProcessLib/Output/Output.h]: > Still, there are unrelated changes in the comments.

These were changed for correcting doxygen error from master. I saw that the doxygen errors  in master have been corrected now. I just removed the changes in documentation including another in
```
MathLib::Point3d getBulkElementPoint(MeshLib::Quad const& quad,
```

File: BaseLib/IO/XmlIO/Qt/XMLQtInterface.h [hunk 7]
```diff
@@ -40,17 +40,17 @@ class XMLQtInterface
     int isValid() const;
 
     /// Checks if a hash for the given data file exists to skip the time-consuming validation part.
-    /// If a hash file exists _and_ the hash of the data file is the same as the content of the hash file the validation is skipped
```
Comment (endJunction) [inline BaseLib/IO/XmlIO/Qt/XMLQtInterface.h:43]: Still, there are unrelated changes in the comments

**Comment by TomFischer** [comment]
Why don't you make the changes per library? This would cause less conflicts with other PRs, the review would be more manageable and therefore much faster?

File: BaseLib/IO/XmlIO/Qt/XMLQtInterface.h (hunk 7 above)
Comment (wenqing) [reply BaseLib/IO/XmlIO/Qt/XMLQtInterface.h]: I only find this one.

**Comment by wenqing** [comment]
> Why don't you make the changes per library? This would cause less conflicts with other PRs, the review would be more manageable and therefore much faster?

I have the changes by library. For the conflicts, they appear in the corresponding  commits of libraries, and they can resolved easily by:
1. applying changes from the base to the remote.
2. copying all from the remote to the base.

Any way, I can split the commits to several PRs. @endJunction and @TomFischer   Which libraries have you reviewed? If you like, I can first put the corresponding commits to another PR for merge.

File: ChemistryLib/PhreeqcIOData/Output.h [hunk 8]
```diff
@@ -24,7 +24,7 @@ class BasicOutputSetups
 public:
     explicit BasicOutputSetups(std::string const& project_file_name,
                                bool const use_high_precision_)
-        : output_file(project_file_name + "phreeqc_.out"),
```
Comment (endJunction) [inline ChemistryLib/PhreeqcIOData/Output.h]: And one more time. Strings must not be changed.

Comment (wenqing) [reply ChemistryLib/PhreeqcIOData/Output.h]: It was corrected but again the change back was committed to the commit for cpp files of ChemistryLib. I will move the correction to the commit for head files of ChemistryLib.

**Comment by TomFischer** [comment]
Please make for every library (GeoLib, BaseLib, MeshLib, ...) a separate PR. Since there might be cases where your scripts could lead to suboptimal changes a review is necessary. If the changes are split per library the review work can be distributed to more developers more easily.

File: Tests/MathLib/TestGaussLegendreIntegration.cpp [hunk 9]
```diff
@@ -276,7 +276,7 @@ struct F3DSeparablePolynomial final : FBase
         return res;
     }
 
-    // [ F(x, y, z) ]_a^b = [ G(x) ]_a^b * [ H(y) ]_a^b * [ I(z) ]_a^b
```
Comment (wenqing) [inline Tests/MathLib/TestGaussLegendreIntegration.cpp:279]: This is missing to correct due to `//`, and it will be corrected.

**Comment by wenqing** [comment]
> Please make for every library (GeoLib, BaseLib, MeshLib, ...) a separate PR. Since there might be cases where your scripts could lead to suboptimal changes a review is necessary. If the changes are split per library the review work can be distributed to more developers more easily.

The thing is that some private members are accessed by the inherited classed in different Libs. I can only splitting this PR into several parts (four or six) that can be compiled, e.g. Application, Base, Mesh and Math in one PR.

**Comment by wenqing** [comment]
or I stop push possible  correction to this PR and wait all reviews are done. The review can be distributed to different reviewers by the commits of Libs.

**Comment by TomFischer** [comment]
> The thing is that some private members are accessed by the inherited classes in different Libs. I can only splitting this PR into several parts (four or six) that can be compiled, e.g. Application, Base, Mesh and Math in one PR.

It would be helpful to divide it into separate parts as much as possible. Then the review can also be distributed more easily, i.e. without great organizational effort, to more developers.

**Comment by wenqing** [comment]
> > The thing is that some private members are accessed by the inherited classes in different Libs. I can only splitting this PR into several parts (four or six) that can be compiled, e.g. Application, Base, Mesh and Math in one PR.
> 
> It would be helpful to divide it into separate parts as much as possible. Then the review can also be distributed more easily, i.e. without great organizational effort, to more developers.

See PR #2980

**Comment by wenqing** [comment]
**How I apply and check the changes**

* Applying changes
  use simple `sed` command in script `sed -i -r 's/(\bb?)_(\w+\b)/\1\2_/g'  [file name]`. This command exactly  moves the underscore of the name of private member to the end of the name (otherwise, the compilation would fail). Therefore one does not need to check the name of private member. However, the command moves the underscore in string (in quotation), macros, comments as well. For those changes, a carefully check is required.

* Check the unnecessary changes
     1 use `grep` command to find string (in quotation), macros, comments in the source code of the  branch of this PR and ufz/master,  and dump the findings into individual files. The commands are
 ```
grep -R '#ifdef' -i -r --include \*.h  --include \*.cpp  &> ifdef.log
grep -R '#ifndef' -i -r --include \*.h  --include \*.cpp  &> ifndef.log
grep -R '#if defined' -i -r --include \*.h  --include \*.cpp  &> ifdefined.log
grep -R '//' -i -r --include \*.h  --include \*.cpp  &> dslash.log
grep -R '/\*' -i -r --include \*.h  --include \*.cpp  &> slash_star.log
grep -R '"' -i -r --include \*.h  --include \*.cpp  &> quot.log
```
      2 compare the files from the branch of the PR and ufz/master to find the unnecessary changes. 
I attach these files here for checking.
[SpecialChangeComparison.zip](https://github.com/ufz/ogs/files/4721814/SpecialChangeComparison.zip)
As you can find in the file comparison, the micros stuffs have no problem (ifdef.log, ifndef.log, ifdefined.log from two branches are identical). I also did not find any problem in other changes shown in these files.

**Comment by wenqing** [comment]
Clang-Analyzer failed, and it seems not related to this PR.

**Comment by endJunction** [comment]
> **How I apply and check the changes**

Thanks for that! That would have saved a lot of time.

**Comment by endJunction** [comment]
Superseded by the follow up pull request #2980
